ome_zarr/reader.py

```diff
@@ -0,0 +1,379 @@
+"""Reading logic for ome-zarr."""
+
+import logging
+from abc import ABC
+from typing import Any, Dict, Iterator, List, Optional, Type, Union, cast
+
+import dask.array as da
+from vispy.color import Colormap
+
+from .conversions import int_to_rgba
+from .io import BaseZarrLocation
+from .types import JSONDict
+
+LOGGER = logging.getLogger("ome_zarr.reader")
+
+
+class Node:
+    """Container for a representation of the binary data somewhere in the data
+    hierarchy."""
+
+    def __init__(
+        self,
+        zarr: BaseZarrLocation,
+        root: Union["Node", "Reader", List[str]],
+        visibility: bool = True,
+    ):
+        self.zarr = zarr
+        self.root = root
+        self.seen: List[str] = []
+        if isinstance(root, Node) or isinstance(root, Reader):
+            self.seen = root.seen
+        else:
+            self.seen = cast(List[str], root)
+        self.__visible = visibility
+
+        # Likely to be updated by specs
+        self.metadata: JSONDict = dict()
+        self.data: List[da.core.Array] = list()
+        self.specs: List[Spec] = []
+        self.pre_nodes: List[Node] = []
+        self.post_nodes: List[Node] = []
+
+        # TODO: this should be some form of plugin infra over subclasses
```

Comment:
I assume another todo might be to add some validation for incompatible specs i.e. I assume a node cannot be both a `Label` and `Labels`

Reply:
If the `Labels` pointed to itself with `.` it might just work...